### ISSUE TITLE
AE_AES_128_CBC_HMAC_SHA_256 and AE_AES_192_CBC_HMAC_SHA_384 functions

### ISSUE THREAD

**Issue by TheDestinyGroup** [alecgn/crypthash-net#3]
Alessandro,

Are there plans to replace cryptSalt and authSalt with salt and derivedKey in the AE_AES_128_CBC_HMAC_SHA_256 and AE_AES_192_CBC_HMAC_SHA_384 functions like you have did to the AE_AES_256_CBC_HMAC_SHA_512 functions a few days ago?

If so is there a timeline for this?

I assume a future release with these changes would result in breaking a change from the 2.4.0 release as the values of EncryptedDataBytes and EncryptedDataBase64String in aesEncryptionResult would have changed from the 2.4.0 release.

Whilst I think this change is a good move I would have concerns about you making such breaking changes going forward if I was to start using this library.

Les

**Comment by alecgn** [comment]
Hi @TheDestinyGroup!

You're right, the latest changes I've made will break the current 2.4.0 version, so I'm gonna release the 3.0 version still this week, if you could, wait for that version. I was planning to make some changes like put the IV and salt int the beggining of the data and the tag in the end (currently IV, salt and tag are put at the end of the data), but actually this order doesn't matters, so I'll keep that as it is. Note that I changed the salt size for AE_AES_192_CBC_HMAC_SHA_384.cs and AE_AES_256_CBC_HMAC_SHA_512.cs, these classes used to have salt with more bits, but 128 bits are really enough.

**Comment by TheDestinyGroup** [comment]
Alessandro,

I'm happy to wait for the 3.0 release before I start using the project.

I see that you've already changed the AE_AES_128_CBC_HMAC_SHA_256 and AE_AES_192_CBC_HMAC_SHA_384 functions. I had missed that being in the same commit as when you changed the AE_AES_256_CBC_HMAC_SHA_512 function as I was focused on looking at the AE_AES_256_CBC_HMAC_SHA_512 EncryptString and DecryptString functions and functions they called.

Whilst I personally think the breaking change is a move in the right direction for the library, I'm certain users of the current release will have a real issue with breaking changes such as this because you don't seem to have provided any backward compatibility for the current method of appending the encryption data so those users who only stored the values of EncryptedDataBytes or EncryptedDataBase64String from aesEncryptionResult when using the current release won't be able to decrypt that data with the next release which will be a major issue for said users.

I know if I was already using this library in a production environment and you made such a breaking change without any providing some form of backward compatibility it would make me hesitant to upgrade to the new release and may even make me reconsider my ongoing use of the library hence why the questions before using this library.

Depending on the number of users you have using the current version in a production environment maybe you might need to consider providing some functions to support converting data with appended encryption data which was encrypted with the current release to the format used by of the new release.

I had noticed the change in the salt lengths when reviewing your recent commits which also contributes to the breaking change for users of the current release who only stored the values of EncryptedDataBytes or EncryptedDataBase64String from aesEncryptionResult when using the current release.

I don't see how putting the IV and salt at the start whilst keeping the tag at the end offers any benefit and makes the breaking change even more breaking for users of earlier releases.

Les

PS: On a side note whilst in no way a bug, but something my OCD nature picked up on when reviewing your AE_AES_256_CBC_HMAC_SHA_512 EncryptString and DecryptString functions and functions they called was the order in which you referenced some of the length variables for the encryption data.

For example you used (encryptedStringBytes.Length - _saltBytesLength  - _IVBytesLength - _tagBytesLength) which is operational correct, but I would have preferred (encryptedStringBytes.Length  - _tagBytesLength- _saltBytesLength  - _IVBytesLength) as the later example reverses the order the in which the encryption data was added and as such makes the code slightly easier to follow.

**Comment by alecgn** [comment]
Hi Les.

Per semantic versioning, I'm gonna increase the major version number to 3 just because I've made incompatible API changes; in despite of having active users of the library (NuGet download numbers show me that), I haven't received any feedback till yours, and I have already made breaking changes when changed from version 1.x.x to 2.0.0; I hope current users are aware of this change of version.
I'm satisfied with the structure of current implementation, and I can guarent that no more breaking changes will be made from version 3 to greater versions, I believe I will stay in this major number, just incrementing the minor and patch numbers. However, if I receive a request from an active user for implement a backward compatibility, certainly I will do it asap.
I'm writing some tests to ensure the quality and the good code bahavior, soon I wish I can coverage all the main functions at least.
About the variable order, I have this kind of OCG when programming too, for me good code seems like poetry... I'm sure it went unnoticed by me while coding, I'll fix that to tranquilize our OCD ok! ;D

**Comment by TheDestinyGroup** [comment]
Alessandro,

Once again thanks for taking the feedback on board.

That seems a reasonable approach.

Whilst there appeared to be around 600 NuGet downloads when I looked at the library I don't know whether that reflects total downloads or downloads of the current release and there is now way of knowing how many of those are just evaluating your library like I am.

I must admit I was surprised there had been no issues (especially comments and questions) raised before I raised my own.

I look forward to the 3.0 release so we can start using the library.

Les
